src/irr/asset/CGLILoader.cpp

```diff
@@ -0,0 +1,157 @@
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "CBufferLoaderBIN.h"
+
+//#ifdef _IRR_COMPILE_WITH_GLI_
+#include "gli/gli.hpp"
+//#else
+//	#error "Need GLI"
+//#endif
+
+namespace irr
+{
+	namespace asset
+	{
+		asset::SAssetBundle CBufferLoaderBIN::loadAsset(io::IReadFile* _file, const asset::IAssetLoader::SAssetLoadParams& _params, asset::IAssetLoader::IAssetLoaderOverride* _override, uint32_t _hierarchyLevel)
+		{
+			if (!_file)
+				return {};
+
+			SContext ctx(_file->getSize());
+			ctx.file = _file;
+
+			const char* filename = _file->getFileName().c_str();
+			gli::texture texture = gli::load(filename);
+			if (texture.empty())
+			{
+				os::Printer::log("Failed to load texture at path ", filename, ELL_ERROR);
+				return {};
+			}
+			
+		    gli::gl GL(gli::gl::PROFILE_GL33);
+			gli::gl::format const format = GL.translate(texture.format(), texture.swizzles());
+			GLenum target = GL.translate(texture.target());
+			
+			/*GLuint textureName = 0;
+			glGenTextures(1, &textureName);
+			glBindTexture(target, textureName);
+			glTexParameteri(target, GL_TEXTURE_BASE_LEVEL, 0);
+			glTexParameteri(target, GL_TEXTURE_MAX_LEVEL, static_cast<GLint>(Texture.levels() - 1));
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_R, format.Swizzles[0]);
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_G, format.Swizzles[1]);
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_B, format.Swizzles[2]);
+			glTexParameteri(target, GL_TEXTURE_SWIZZLE_A, format.Swizzles[3]);*/
+			
+			glm::tvec3<GLsizei> const extent(texture.extent()); // it is on 0 level without specyfing it explicitly
+			GLsizei const totalFaces = static_cast<GLsizei>(texture.layers() * texture.faces());
+
+			ICPUImage::SCreationParams imageInfo;
+			imageInfo.format = format;
+			imageInfo.extent.width = extent.x;
+			imageInfo.extent.height = extent.y;
+			imageInfo.extent.depth = extent.z;
+			imageInfo.mipLevels = texture.levels();
+			imageInfo.arrayLayers = texture.layers();
```

Comment:
total faces

Reply:
so what is a difference between layer and face?

Reply:
that's a distinction that only GLi seems to make, in Vulkan and modern (DSA) OpenGL, faces of a cubemap are layers 

`arrayLayerIndex%6` gives you the cubemap face
`arrayLayerIndex/6` gives you the cubemap array index

Reply:
hm, actually it seems to me that our image assumes that the "layer" is always one, but if it is an cubemap it can handle some "layers" considered by our API as faces, right?
so if there were some layers, I have to create pointers of our **ICPUImage** and if target is an cubemap, I can set **arrayLayers** that stores our faces

Reply:
but in that case where there are loaded layers I have to consider mipmap level if present and create **SCreationParams** for each image

Reply:
only one ICPUImage is ever produced, "layers" are no different than "height" pixel addressing.

ICPUImage consists of declaration of image size+format+etc plus a list of buffer regions that shall be copied into it, see the "fixed" PNG loader, separate mip-map levels are separate image subresources, hence a separate buffer-region to copy. You don't however need 1 buffer per copy region, you can pack everything into one without any problem (that's what offsets are for).

See Vulkan-Tutorial and read documentation of `vkCmdCopyBufferToImage`